src/tools/rbd_mirror/PoolWatcher.cc

```diff
@@ -115,7 +116,13 @@ void PoolWatcher::refresh_images(bool reschedule)
         continue;
       }
       for (auto it = mirror_images.begin(); it != mirror_images.end(); ++it) {
-        image_ids.insert(ImageIds(it->first, it->second));
+        std::string image_name;
+        r = dir_get_name(&ioctx, RBD_DIRECTORY, it->first, &image_name);
```

Comment:
I would recommend a single pass of retrieving all image names in the pool, followed by retrieving the mirrored images, and then cross-reference.  I'm not sure we can skip adding the mirrored image to the collection since that would be interpreted as a deleted image propagation.  Maybe a boost::optional<std::string> for the image name?